sandbox/drone-imagery-ingestion/07_extract-exif-metadata-sub-mission.R

```diff
@@ -25,59 +28,74 @@ crosswalk_filepath = file.path(FOLDER_BASEROW_CROSSWALK_PATH, paste0(IMAGERY_PRO
+exif = exif |> mutate(mission_id = str_pad(mission_id, 6, pad = "0", side = "left"))
 
-# Extract image-level metadata, which can occur across all missions at once becuase there are no
-# hierarchical dependencies on mission-level data
-metadata_perimage = extract_imagery_perimage_metadata(exif,
-                                                      input_type = "dataframe")
-
-# Assign image_id to the exif dataframe so it can be used in the dataset-level metadata extraction
-exif$image_id = metadata_perimage$image_id
-
-# For sub-mission-level metadata, run sub-mission by sub-mission
-submissions = unique(exif$submission_id)
+# Compute the unique sub mission IDs
+unique_sub_missions = unique(exif$submission_id)
 
 # For parallelizing, make a list of subsets of the exif dataframe, one for each sub-mission
-exif_list <- lapply(submissions, function(submission) {
-  exif_foc <- exif |>
-    filter(submission_id == submission)
-  return(exif_foc)
+exif_per_sub_mission <- lapply(
+  unique_sub_missions,
+  function(sub_mission_ID) {
+    # Extract the exif rows matching that sub-mission ID
+    sub_mission_exif <- exif |>
+      filter(submission_id == sub_mission_ID)
+    return(sub_mission_exif)
+  }
+)
+
+# Compute the corresponding baserow data for each of the sub-missions
+aircraft_model_names <- lapply(
+  unique_sub_missions,
+  function(sub_mission_ID) {
+    # Extract the baserow entries for this sub-mission ID
+    baserow_for_sub_mission <- baserow_sub_mission_metadata[
+      baserow_sub_mission_metadata$sub_mission_id == sub_mission_ID,
+    ]
+
+    # There should only be one matching row
+    if (nrow(baserow_for_sub_mission) != 1) {
+      stop(paste("Error: there was not one corresponding baserow entry: ", baserow_for_sub_mission))
+    }
+
+    # Extract the aircraft model name
+    aircraft_model_name = baserow_for_sub_mission$aircraft_model_name
+
+    return(aircraft_model_name)
+  }
+)
+
+# Extract the metadata in a standardized manner no matter the platform
+metadata_per_sub_dataset = furrr::future_map(seq_along(exif_per_sub_mission), function(i) {
+  extract_imagery_perimage_metadata(exif = exif_per_sub_mission[[i]], platform_name = aircraft_model_names[[i]])
 })
+print("Finished processing per-sub-mission datasets")
 
-# Run dataset-level metadata extraction across each subset
-
-future::plan("multisession")
-res = furrr::future_map(exif_list,
-                        extract_imagery_dataset_metadata,
-                        input_type = "dataframe",
-                        plot_flightpath = FALSE,
-                        crop_to_contiguous = TRUE,
-                        min_contig_area = 1600,
-                        .options = furrr_options(seed = TRUE))
-
-metadata_list = map(res, ~.x$dataset_metadata)
-polygon_list = map(res, ~.x$mission_polygon)
-images_retained_list = map(res, ~.x$images_retained)
+summary_statistics = furrr::future_map(metadata_per_sub_dataset, extract_imagery_dataset_metadata)
```

Comment:
However in testing I noticed that (only when run with parallelization there is an error like `Cannot find function st_zm` which is really confusing me because every call to st_zm in the `ofo-r` codebase is prefaced correctly with `sf::`. So it must be a third-party function calling st_zm but it will be really hard to debug because it only happens with parallelization enabled.

Reply:
Hmm, that is weird. Do you think multiple threads can download the DTM at the same rate as only one thread, or would they be fighting for network bandwidth?

Reply:
There was a single instance of `st_zm` that wasn't prefixed in `R/tree-detection-accuracy-assessment.R`. I fixed that and the error went away.

Reply:
Ah good find on the one `st_zm`. I think multiple threads (up to some number like 8?) will each have the download speed of a single serial thread, so parallelizing should speed it up substantially. Also there appears to be a lot of per-DEM overhead (not sure what -- waiting for the API to respond? -- that is not bandwidth-related and should scale pretty linearly with number of threads.